### ISSUE TITLE
ishft much slower than in IDL

### ISSUE THREAD

**Issue by brandy125** [gnudatalanguage/gdl#922]
I have many millions of calls to ishft.
The implementation in GDL is around 50 times slower than in IDL:

GDL> t=systime(1) & for i=0l,1d6 do tmp=(ishft(0,0) ) & print,systime(1)-t
       2.9121642

IDL>  t=systime(1) & for i=0l,1d6 do tmp=(ishft(0,0) ) & print,systime(1)-t
     0.053571939

Is there any chance to speed this up?

**Comment by GillesDuvert** [comment]
Hi @brandy125 you have found the worst speed example.
```
GDL> t=systime(1) & for i=0l,1d6 do tmp=(ishft(0,0) ) & print,systime(1)-t
       2.4089401
```
if I do
```
GDL> a=0
GDL> t=systime(1) & for i=0l,1d6 do tmp=(ishft(a,0) ) & print,systime(1)-t
       1.7602870
```
meaning that creating one million times an internal variable from the 0 instead of an existing variable (a) takes already 0.7 seconds.

On a real-size example:
```
GDL> a=fix(dist(30))
GDL> t=systime(1) & for i=0l,1d6 do tmp=(ishft(a,0) ) & print,systime(1)-t
       3.3742149
```
```
IDL>  a=fix(dist(30))                                                     
IDL> t=systime(1) & for i=0l,1d6 do tmp=(ishft(a,0) ) & print,systime(1)-t
       1.3758190
```
GDL is not so bad.

So

- yes GDL is slower in loops, and definitely slow in creating variables (I know why)
- It's probably easy to write a special fast case for small sized problems, this is why I tag this request.
- the zero shift doing nothing, it could be avoided in the GDL code, but also in your procedure?

**Comment by brandy125** [comment]
Hi @GillesDuvert, thank you for your analysis. 

You are right that IDL gets slow when working on an array of numbers and GDL gets compatible.
In my special case I need to apply a shift to a variable of one element only and here IDL is 50 times faster than GDL. 
This is a real example of what I do (with little different numbers and little different shifts but this does not matter here):

```
GDL> t=systime(1) & cmd=188878546031083560LL & for i=0l,1d6 do tmp=ishft(cmd,-1L) & print,(systime(1)-t)
           2.4174190
```
while in IDL:

```
IDL> t=systime(1) & cmd=188878546031083560LL & for i=0l,1d6 do tmp=ishft(cmd,-1L) & print,(systime(1)-t)
         0.059509039 
```

The point for me is that IDL is so bloody fast if the shift is applied to a single number.

**Comment by GillesDuvert** [comment]
note that GDL is inherently slow on loops, and probably will stay so.
```
GDL> t=systime(1) & cmd=188878546031083560LL & for i=0l,1d6 do wait,0 & print,(systime(1)-t)
      0.75977898
```

**Comment by alaingdl** [comment]
test_ishft is not OK with the last changes

**Comment by GillesDuvert** [comment]
Please Alain be more precise, test_ishft is OK on my machine.

**Comment by GillesDuvert** [comment]
closed by #937 .
I forgot in my previous comment that I was running a new unpublished version without the problem.
